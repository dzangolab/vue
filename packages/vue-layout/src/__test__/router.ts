import { createRouter, createWebHistory } from "vue-router";

const Test = {
  name: "Test",
  template: `<h1>Test</h1>`,
};

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      component: Test,
      path: "/",
    },
  ],
});

export default router;
