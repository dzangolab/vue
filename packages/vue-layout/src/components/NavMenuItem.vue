<template>
  <div class="nav-menu-item">
    <a class="link" aria-label="open menu" @click="onClick">
      <i v-if="item.icon" :class="`icon ${item.icon}`" />
      <span v-if="showShortName">{{ item.shortName }}</span>
      <span class="item-name" :class="sidebarActive ? 'active-item' : ''">
        {{ item.name }}
      </span>
    </a>
    <div
      v-if="item.children && item.children.length && showChildren"
      class="sub-menu-item"
    >
      <NavMenuItem
        v-for="child in item.children"
        :key="child.name"
        :item="child"
        :sidebar-active="sidebarActive"
      />
    </div>
  </div>
</template>

<script lang="ts">
export default {
  name: "NavMenuItem",

  components: {
    NavMenuItem: () => import("./NavMenuItem.vue"),
  },
};
</script>

<script setup lang="ts">
import { computed, ref } from "vue";
import { useRouter } from "vue-router";

const props = defineProps({
  item: {
    required: true,
    type: Object,
  },

  sidebarActive: {
    default: true,
    required: true,
    type: Boolean,
  },
});

const router = useRouter();

const showChildren = ref<boolean>(false);

const showShortName = computed(() => {
  return !props.sidebarActive && props.item.shortName;
});

const onClick = () => {
  if (props.item.routeName) {
    router.push({ name: props.item.routeName });
  } else {
    showChildren.value = !showChildren.value;
  }
};
</script>

<style lang="css">
.link {
  align-items: center;
  color: #636363;
  cursor: pointer;
  display: flex;
  font-size: 12px;
  font-weight: 450;
  height: 49.19px;
  padding-left: 12px;
  text-decoration: none;
  text-transform: uppercase;
  transition: 0.5s;
}

.sub-menu-item {
  margin-left: 2rem;
}
</style>
