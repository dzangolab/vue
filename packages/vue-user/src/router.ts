import { storeToRefs } from "pinia";
import { Router } from "vue-router";

import AuthGoogleCallback from "./components/AuthGoogleCallback.vue";
import useUserStore from "./store";
import { getUser } from "./supertokens";
import VerifyEmailReminder from "./views/EmailVerificationReminder.vue";
import Login from "./views/Login.vue";
import PasswordReset from "./views/PasswordReset.vue";
import PasswordResetRequest from "./views/PasswordResetRequest.vue";
import PasswordResetRequestAcknowledge from "./views/PasswordResetRequestAcknowledge.vue";
import Profile from "./views/Profile.vue";
import Signup from "./views/Signup.vue";

import type { RouteOverride, RouteOverrides } from "./types";
import type { RouteMeta, RouteRecordRaw } from "vue-router";

const _routes = {
  google: {
    component: AuthGoogleCallback,
    name: "authcallback",
    path: "/auth/callback/google",
  },
  login: {
    component: Login,
    name: "login",
    path: "/login",
  } as RouteRecordRaw,
  profile: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: Profile,
    name: "profile",
    path: "/profile",
  } as RouteRecordRaw,
  signup: {
    component: Signup,
    name: "signup",
    path: "/signup",
  } as RouteRecordRaw,
  passwordReset: {
    component: PasswordReset,
    name: "resetPassword",
    path: "/reset-password/:token?",
  } as RouteRecordRaw,
  passwordResetRequest: {
    component: PasswordResetRequest,
    name: "resetPasswordRequest",
    path: "/reset-password-request",
  } as RouteRecordRaw,
  passwordResetRequestAcknowledge: {
    component: PasswordResetRequestAcknowledge,
    name: "resetPasswordRequestAcknowledge",
    path: "/reset-password-request-acknowledge",
  },
  verifyEmailReminder: {
    meta: {
      authenticated: true,
    } as RouteMeta,
    component: VerifyEmailReminder,
    name: "verifyEmailReminder",
    path: "/verify-email-reminder",
  },
};

const getRoute = (
  defaultRoute: object,
  override?: RouteOverride,
): RouteRecordRaw => {
  return {
    ...defaultRoute,
    ...override,
  } as RouteRecordRaw;
};

const addRoutes = (router: Router, routes?: RouteOverrides) => {
  router.addRoute(getRoute(_routes.google, routes?.google));

  router.addRoute(getRoute(_routes.login, routes?.login));

  if (!routes?.signup?.disabled) {
    router.addRoute(getRoute(_routes.signup, routes?.signup));
  }

  router.addRoute(getRoute(_routes.passwordReset, routes?.passwordReset));

  router.addRoute(
    getRoute(_routes.passwordResetRequest, routes?.passwordResetRequest),
  );

  router.addRoute(getRoute(_routes.profile, routes?.profile));

  router.addRoute(
    getRoute(
      _routes.passwordResetRequestAcknowledge,
      routes?.passwordResetRequestAcknowledge,
    ),
  );

  router.addRoute(
    getRoute(_routes.verifyEmailReminder, routes?.verifyEmailReminder),
  );
};

const redirectRoutes = (router: Router) => {
  router.beforeEach((to, from, next) => {
    const userStore = useUserStore();
    const { user } = storeToRefs(userStore);

    const routesToRedirect = [
      "login",
      "signup",
      "resetPassword",
      "resetPasswordRequest",
    ];
    const name = to.name as string;

    if (user.value && routesToRedirect.includes(name)) {
      next({ name: "profile" });
    } else {
      next();
    }
  });
};

const addAuthenticationGuard = (router: Router) => {
  router.beforeEach(async (to) => {
    const meta = to.meta as RouteMeta;

    const userStore = useUserStore();

    const { isEmailVerified } = storeToRefs(userStore);
    const name = to.name as string;
    const routesToRedirect = ["verifyEmailReminder"];
    const { user } = storeToRefs(userStore);

    if (!user.value) {
      user.value = await getUser();
    }
    if (meta.authenticated && !user.value) {
      router.push({ name: "login" }); // using next inside async function is not allowed
    } else if (meta.authenticated && user.value && !isEmailVerified.value) {
      router.push({ name: "verifyEmailReminder" });
    } else if (
      user.value &&
      isEmailVerified.value &&
      routesToRedirect.includes(name)
    ) {
      router.push({ name: "home" });
    }
  });
};

const updateRouter = (router: Router, routes?: RouteOverrides) => {
  addRoutes(router, routes);

  addAuthenticationGuard(router);

  redirectRoutes(router);
};

export default updateRouter;
