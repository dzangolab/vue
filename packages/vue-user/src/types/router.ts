import type { RouteComponent } from "vue-router";

interface RouteOverride {
  component?: RouteComponent;
  path?: string;
}

interface RouteOverrides {
  google?: RouteOverride;
  home?: string;
  login: RouteOverride;
  signup: RouteOverride & { disabled?: boolean };
  passwordReset: RouteOverride;
  passwordResetRequest: RouteOverride;
  passwordResetRequestAcknowledge: RouteOverride;
  profile?: RouteOverride;
}

export type { RouteOverride, RouteOverrides };
