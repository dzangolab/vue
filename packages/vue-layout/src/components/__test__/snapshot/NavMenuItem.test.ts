import { mount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";

import type { VueWrapper } from "@vue/test-utils";

import router from "@/__test__/router";
import NavMenuItem from "@/components/NavMenuItem.vue";

describe("NavMenuItem", () => {
  const wrapper: VueWrapper = mount(NavMenuItem, {
    global: {
      mocks: {
        showChildren: true,
      },
      plugins: [router],
    },
    props: {
      item: {
        name: "Menu1",
        routeName: "menu-1",
        shortName: "M1",
        children: [
          {
            name: "child1",
            routeName: "child-1",
            shortName: "C1",
          },
          {
            name: "child2",
            routeName: "child-2",
            shortName: "C2",
          },
        ],
      },
      sidebarActive: true,
    },
  });

  it("matches snapshot ", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
