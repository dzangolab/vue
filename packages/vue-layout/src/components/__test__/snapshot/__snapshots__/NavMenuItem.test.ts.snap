// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavMenuItem > matches snapshot  1`] = `
<div
  class="nav-menu-item"
>
  <a
    aria-label="open menu"
    class="link"
  >
    <!--v-if-->
    <!--v-if-->
    <span
      class="item-name"
    >
      Menu1
    </span>
    <span
      class="toggle-menu"
    >
      <svg
        class="up-chevron"
        fill="none"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.29289 8.29289C4.68342 7.90237 5.31658 7.90237 5.70711 8.29289L12 14.5858L18.2929 8.29289C18.6834 7.90237 19.3166 7.90237 19.7071 8.29289C20.0976 8.68342 20.0976 9.31658 19.7071 9.70711L12.7071 16.7071C12.3166 17.0976 11.6834 17.0976 11.2929 16.7071L4.29289 9.70711C3.90237 9.31658 3.90237 8.68342 4.29289 8.29289Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </a>
  <transition-stub
    appear="false"
    css="true"
    name="fade"
    persisted="false"
  >
    <div
      class="sub-menu-item"
    >
      
      <div
        class="nav-menu-item"
      >
        <a
          aria-label="open menu"
          class="link"
        >
          <!--v-if-->
          <!--v-if-->
          <span
            class="item-name"
          >
            child1
          </span>
          <!--v-if-->
        </a>
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </div>
      <div
        class="nav-menu-item"
      >
        <a
          aria-label="open menu"
          class="link"
        >
          <!--v-if-->
          <!--v-if-->
          <span
            class="item-name"
          >
            child2
          </span>
          <!--v-if-->
        </a>
        <transition-stub
          appear="false"
          css="true"
          name="fade"
          persisted="false"
        >
          <!--v-if-->
        </transition-stub>
      </div>
      
    </div>
  </transition-stub>
</div>
`;
